File: Musicefy.Core/Services/YouTube/YouTubeUrlParser.cs
using System;
using System.Text.RegularExpressions;

namespace Musicefy.Core.Services.YouTubeApi
{
    /// <summary>
    /// Parser for various YouTube URL formats.
    /// Inspired by Echo Music's YouTubeUrlParser that handles watch, shorts, embed, and music URLs.
    /// </summary>
    public static class YouTubeUrlParser
    {
        // Video URL patterns (inspired by Echo Music's VIDEO_URL_PATTERNS)
        private static readonly Regex[] VideoUrlPatterns = new[]
        {
            // Standard watch URL
            new Regex(@"(?:https?://)?(?:www\.)?(?:music\.)?youtube\.com/watch\?.*v=([a-zA-Z0-9_-]{11})", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            // Short URL
            new Regex(@"(?:https?://)?youtu\.be/([a-zA-Z0-9_-]{11})", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            // Shorts URL
            new Regex(@"(?:https?://)?(?:www\.)?youtube\.com/shorts/([a-zA-Z0-9_-]{11})", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            // Embed URL
            new Regex(@"(?:https?://)?(?:www\.)?youtube\.com/embed/([a-zA-Z0-9_-]{11})", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            // Music watch URL
            new Regex(@"(?:https?://)?music\.youtube\.com/watch\?.*v=([a-zA-Z0-9_-]{11})", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        };

        // Playlist URL patterns
        private static readonly Regex[] PlaylistUrlPatterns = new[]
        {
            new Regex(@"(?:https?://)?(?:www\.)?(?:music\.)?youtube\.com/playlist\?.*list=([a-zA-Z0-9_-]+)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            new Regex(@"(?:https?://)?(?:www\.)?(?:music\.)?youtube\.com/watch\?.*list=([a-zA-Z0-9_-]+)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        };

        // Artist/channel URL patterns (inspired by Echo Music's ARTIST_URL_PATTERNS)
        private static readonly Regex[] ArtistUrlPatterns = new[]
        {
            new Regex(@"(?:https?://)?(?:www\.)?music\.youtube\.com/channel/(UC[a-zA-Z0-9_-]+)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
            new Regex(@"(?:https?://)?(?:www\.)?music\.youtube\.com/browse/(MPRE[a-zA-Z0-9_-]+)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        };

        /// <summary>
        /// Parsed result of a YouTube URL.
        /// </summary>
        public class ParsedYouTubeUrl
        {
            public string VideoId { get; set; }
            public string PlaylistId { get; set; }
            public string BrowseId { get; set; }
            public UrlType Type { get; set; }
        }

        public enum UrlType
        {
            Video,
            Playlist,
            Artist,
            Album,
            Unknown
        }

        /// <summary>
        /// Parse a YouTube URL and extract the video/playlist/artist ID.
        /// </summary>
        public static ParsedYouTubeUrl Parse(string url)
        {
            if (string.IsNullOrEmpty(url))
                return new ParsedYouTubeUrl { Type = UrlType.Unknown };

            // Try video patterns
            foreach (var pattern in VideoUrlPatterns)
            {
                var match = pattern.Match(url);
                if (match.Success)
                {
                    var result = new ParsedYouTubeUrl
                    {
                        VideoId = match.Groups[1].Value,
                        Type = UrlType.Video
                    };

                    // Also check for playlist ID in the same URL
                    foreach (var plPattern in PlaylistUrlPatterns)
                    {
                        var plMatch = plPattern.Match(url);
                        if (plMatch.Success)
                        {
                            result.PlaylistId = plMatch.Groups[1].Value;
                            break;
                        }
                    }

                    return result;
                }
            }

            // Try playlist patterns
            foreach (var pattern in PlaylistUrlPatterns)
            {
                var match = pattern.Match(url);
                if (match.Success)
                    return new ParsedYouTubeUrl { PlaylistId = match.Groups[1].Value, Type = UrlType.Playlist };
            }

            // Try artist/channel patterns
            foreach (var pattern in ArtistUrlPatterns)
            {
                var match = pattern.Match(url);
                if (match.Success)
                {
                    var id = match.Groups[1].Value;
                    return new ParsedYouTubeUrl
                    {
                        BrowseId = id,
                        Type = id.StartsWith("MPRE") ? UrlType.Album : UrlType.Artist
                    };
                }
            }

            return new ParsedYouTubeUrl { Type = UrlType.Unknown };
        }

        /// <summary>
        /// Extract just the video ID from a URL or ID string.
        /// </summary>
        public static string ExtractVideoId(string urlOrId)
        {
            if (string.IsNullOrEmpty(urlOrId))
                return null;

            // If it's already a valid video ID (11 chars, alphanumeric + - _)
            if (IsValidVideoId(urlOrId))
                return urlOrId;

            // Try parsing as URL
            var parsed = Parse(urlOrId);
            return parsed.VideoId;
        }

        /// <summary>
        /// Check if a string is a valid YouTube video ID.
        /// </summary>
        public static bool IsValidVideoId(string id)
        {
            if (string.IsNullOrEmpty(id) || id.Length != 11)
                return false;

            foreach (char c in id)
            {
                if (!char.IsLetterOrDigit(c) && c != '-' && c != '_')
                    return false;
            }

            return true;
        }

        /// <summary>
        /// Create a YouTube Music watch URL from a video ID.
        /// </summary>
        public static string CreateWatchUrl(string videoId)
        {
            return $"https://music.youtube.com/watch?v={videoId}";
        }
    }
}

File: src/core/sources/youtube/YouTubeUrlParser.cpp
// YouTubeUrlParser.cpp

#include "YouTubeUrlParser.h"

#include <QRegularExpression>

namespace mf::core::sources::youtube {

// ── Video ID patterns ─────────────────────────────────────────────
// All capture exactly 11-char video IDs: [a-zA-Z0-9_-]{11}
static const QRegularExpression kVideoPatterns[] = {
    // youtube.com/watch?...v=XX
    QRegularExpression(R"(youtube\.com/.*[?&]v=([a-zA-Z0-9_-]{11}))"),
    // youtu.be/XX
    QRegularExpression(R"(youtu\.be/([a-zA-Z0-9_-]{11}))"),
    // youtube.com/shorts/XX
    QRegularExpression(R"(youtube\.com/shorts/([a-zA-Z0-9_-]{11}))"),
    // youtube.com/embed/XX
    QRegularExpression(R"(youtube\.com/embed/([a-zA-Z0-9_-]{11}))"),
    // music.youtube.com/watch?...v=XX
    QRegularExpression(R"(music\.youtube\.com/.*[?&]v=([a-zA-Z0-9_-]{11}))"),
};

// ── Playlist ID patterns ──────────────────────────────────────────
static const QRegularExpression kPlaylistPatterns[] = {
    // youtube.com/playlist?...list=XX
    QRegularExpression(R"(youtube\.com/playlist\?.*[?&]list=([a-zA-Z0-9_-]+))"),
    // youtube.com/watch?...list=XX  (video URL with playlist parameter)
    QRegularExpression(R"(youtube\.com/.*[?&]list=([a-zA-Z0-9_-]+))"),
};

// ── Artist (channel) pattern ──────────────────────────────────────
// music.youtube.com/channel/UCXX
static const QRegularExpression kArtistPattern(
    R"(music\.youtube\.com/channel/(UC[a-zA-Z0-9_-]+))");

// ── Album (browse) pattern ────────────────────────────────────────
// music.youtube.com/browse/MPREXX
static const QRegularExpression kAlbumPattern(
    R"(music\.youtube\.com/browse/(MPRE[a-zA-Z0-9_-]+))");

// ──────────────────────────────────────────────────────────────────
ParsedUrl YouTubeUrlParser::parse(const QString& url)
{
    const QString trimmed = url.trimmed();
    ParsedUrl result;

    // Try artist/album first (music.youtube.com specific)
    if (matchArtistId(trimmed, result.browseId)) {
        result.type = UrlType::Artist;
        return result;
    }
    if (matchAlbumId(trimmed, result.browseId)) {
        result.type = UrlType::Album;
        return result;
    }

    // Try video patterns
    if (matchVideoId(trimmed, result.videoId)) {
        result.type = UrlType::Video;
        // Check if the URL also contains a playlist parameter
        for (const auto& pat : kPlaylistPatterns) {
            auto m = pat.match(trimmed);
            if (m.hasMatch()) {
                result.playlistId = m.captured(1);
                break;
            }
        }
        return result;
    }

    // Try playlist patterns
    if (matchPlaylistId(trimmed, result.playlistId)) {
        result.type = UrlType::Playlist;
        return result;
    }

    return result; // UrlType::Unknown
}

// ──────────────────────────────────────────────────────────────────
bool YouTubeUrlParser::isYouTubeUrl(const QString& url)
{
    return parse(url).type != UrlType::Unknown;
}

// ──────────────────────────────────────────────────────────────────
QString YouTubeUrlParser::createWatchUrl(const QString& videoId)
{
    return QStringLiteral("https://music.youtube.com/watch?v=%1").arg(videoId);
}

// ──────────────────────────────────────────────────────────────────
bool YouTubeUrlParser::matchVideoId(const QString& url, QString& out)
{
    for (const auto& pat : kVideoPatterns) {
        auto m = pat.match(url);
        if (m.hasMatch()) {
            out = m.captured(1);
            return true;
        }
    }
    return false;
}

// ──────────────────────────────────────────────────────────────────
bool YouTubeUrlParser::matchPlaylistId(const QString& url, QString& out)
{
    for (const auto& pat : kPlaylistPatterns) {
        auto m = pat.match(url);
        if (m.hasMatch()) {
            out = m.captured(1);
            return true;
        }
    }
    return false;
}

// ──────────────────────────────────────────────────────────────────
bool YouTubeUrlParser::matchArtistId(const QString& url, QString& out)
{
    auto m = kArtistPattern.match(url);
    if (m.hasMatch()) {
        out = m.captured(1);
        return true;
    }
    return false;
}

// ──────────────────────────────────────────────────────────────────
bool YouTubeUrlParser::matchAlbumId(const QString& url, QString& out)
{
    auto m = kAlbumPattern.match(url);
    if (m.hasMatch()) {
        out = m.captured(1);
        return true;
    }
    return false;
}

} // namespace mf::core::sources::youtube
